header/p2p/server.go

```diff
@@ -52,19 +54,66 @@ func (serv *ExchangeServer) Stop(context.Context) error {
 
 // requestHandler handles inbound ExtendedHeaderRequests.
 func (serv *ExchangeServer) requestHandler(stream network.Stream) {
+	err := stream.SetReadDeadline(time.Now().Add(readDeadline))
+	if err != nil {
+		log.Debugf("error setting deadline: %s", err)
+	}
 	// unmarshal request
 	pbreq := new(p2p_pb.ExtendedHeaderRequest)
-	_, err := serde.Read(stream, pbreq)
+	_, err = serde.Read(stream, pbreq)
 	if err != nil {
 		log.Errorw("server: reading header request from stream", "err", err)
 		stream.Reset() //nolint:errcheck
 		return
 	}
+	if err = stream.CloseRead(); err != nil {
+		log.Error(err)
+	}
+
+	var headers []*header.ExtendedHeader
 	// retrieve and write ExtendedHeaders
-	if pbreq.Hash != nil {
-		serv.handleRequestByHash(pbreq.Hash, stream)
-	} else {
-		serv.handleRequest(pbreq.Origin, pbreq.Origin+pbreq.Amount, stream)
+	switch pbreq.Data.(type) {
+	case *p2p_pb.ExtendedHeaderRequest_Hash:
+		headers, err = serv.handleRequestByHash(pbreq.GetHash())
+	case *p2p_pb.ExtendedHeaderRequest_Origin:
+		headers, err = serv.handleRequest(pbreq.GetOrigin(), pbreq.GetOrigin()+pbreq.Amount)
+	default:
+		log.Error("server: invalid data type received")
+		stream.Reset() //nolint:errcheck
+		return
+	}
+	code := p2p_pb.StatusCode_OK
+	if err != nil {
+		switch err {
+		case header.ErrNotFound:
+			headers = make([]*header.ExtendedHeader, 1)
+			code = p2p_pb.StatusCode_NOT_FOUND
+		default:
+			stream.Reset() //nolint:errcheck
+			return
+		}
+	}
```

Comment:
Looks like this part is handling error from `stream.CloseRead()` happening at line 69? Maybe it is possible to combine it with same if err != nil{} check? 

Also optionally `switch` with 1 case could be reduced by 
```
if !errorrs.Is(header.ErrNotFound){ return fast}
```

Reply:
> Looks like this part is handling error from stream.CloseRead()
Not exactly. It handles error from `handleRequestByHash/handleRequest`

> Also optionally switch with 1 case could be reduced by
there will be one more case in subsuquent pr